compiler/rustc_builtin_macros/src/concat_bytes.rs

```diff
@@ -0,0 +1,210 @@
+                                "try using a byte character",
+                                format!("b{}", snippet),
+                                Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    } else {
+                        cx.span_err(e.span, "cannot concatenate character literals")
+                    }
+                    has_errors = true;
+                }
+                ast::LitKind::Int(
+                    val,
+                    ast::LitIntType::Unsuffixed | ast::LitIntType::Unsigned(ast::UintTy::U8),
+                ) if val <= 255 => {
+                    if let Ok(snippet) = cx.sess.source_map().span_to_snippet(e.span) {
+                        cx.struct_span_err(e.span, "cannot concatenate numeric literals")
+                            .span_suggestion(
+                                e.span,
+                                "try wrapping the number in an array",
+                                format!("[{}]", snippet),
+                                Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    } else {
+                        cx.span_err(e.span, "cannot concatenate numeric literals")
+                    }
+                    has_errors = true;
+                }
+                ast::LitKind::Int(_, _) => {
+                    cx.span_err(e.span, "cannot concatenate numeric literals");
+                    has_errors = true;
+                }
+                ast::LitKind::Bool(val) => {
+                    cx.struct_span_err(e.span, "cannot concatenate boolean literals")
+                        .span_suggestion(
+                            e.span,
+                            "try using the numeric value",
+                            (val as u8).to_string(),
+                            Applicability::MachineApplicable,
+                        )
+                        .emit();
+                    has_errors = true;
+                }
+                ast::LitKind::Byte(val) => {
+                    accumulator.push(val);
+                }
+                ast::LitKind::ByteStr(ref bytes) => {
+                    accumulator.extend_from_slice(&bytes);
+                }
+                ast::LitKind::Err(_) => {
+                    has_errors = true;
+                }
+            },
+            ast::ExprKind::Err => {
+                has_errors = true;
+            }
+            _ => {
+                missing_literal.push(e.span);
+            }
+        }
+    }
+    if !missing_literal.is_empty() {
+        let mut err = cx.struct_span_err(missing_literal, "expected a byte literal");
+        err.note("only byte literals (like `b\"foo\"`, `b's'`, and `[3, 4, 5]`) can be passed to `concat_bytes!()`");
+        err.emit();
+        return DummyResult::any(sp);
+    } else if has_errors {
+        return DummyResult::any(sp);
```

Comment:
Hm. I'm not sure if this is better, but maybe we should consider emitting `DummyResult::raw_expr(sp, true)` here? That way we're still indicating this will be an expression...